compiler/src/iree/compiler/Dialect/Flow/TransformExtensions/FlowExtensions.cpp

```diff
@@ -446,5 +447,294 @@ transform_dialect::ForeachThreadToFlowDispatchWorkgroupsOp::applyToOne(
+#ifndef NDEBUG
+      ShapedType tiedOperandType = tiedOperand.getType().cast<ShapedType>();
+      assert(tiedOperandType == shapedType && "expected same type");
+#endif  // NDEBUG
+      return reifyDynamicResultDims(b, tiedOperand, dynamicDims);
+    }
+  }
+
+  // Case 4: Query ShapeAwareOpInterface.
+  auto shapeAwareOp = dyn_cast<IREE::Util::ShapeAwareOpInterface>(op);
+  if (shapeAwareOp) {
+    ValueRange dims =
+        shapeAwareOp.getResultDynamicDims(opResult.getResultNumber());
+    dynamicDims.append(dims.begin(), dims.end());
+    return success();
+  }
+
+  // Case 5: Query ReifyRankedShapedTypeOpInterface.
+  auto reifyShapeOp = dyn_cast<ReifyRankedShapedTypeOpInterface>(op);
+  if (reifyShapeOp) {
+    ReifiedRankedShapedTypeDims dims;
+    if (failed(reifyShapeOp.reifyResultShapes(b, dims))) return failure();
+    for (int64_t i = 0; i < shapedType.getRank(); ++i)
+      if (shapedType.isDynamicDim(i))
+        dynamicDims.push_back(dims[opResult.getResultNumber()][i]);
+    return success();
+  }
+
+  return failure();
+}
+
+/// Reify the dynamic dimensions of all results of the given op.
+static LogicalResult reifyDynamicResultDims(OpBuilder b, Operation *op,
+                                            SmallVector<Value> &dynamicDims) {
+  for (Value v : op->getResults())
+    if (failed(reifyDynamicResultDims(b, v, dynamicDims))) return failure();
+
+  return success();
+}
+
+static LogicalResult moveIntoDispatchRegion(RewriterBase &rewriter,
+                                            Operation *target,
+                                            Flow::DispatchRegionOp &regionOp) {
+  OpBuilder::InsertionGuard guard(rewriter);
+  Block &body = regionOp.getBody().front();
+  DominanceInfo domInfo;
+
+  // Case 1: `target` dominates `regionOp`.
+  //
+  // Example:
+  //
+  // %0 = "some_op"() : () -> (tensor<?xf32>)
+  // %r = flow.dispatch.region -> (tensor<?xf32>{%d0}) {
+  //   %1 = "another_op"(%0) : (tensor<?xf32>) -> (tensor<?xf32>)
+  //   flow.return %1 : tensor<?xf32>
+  // }
+  // %2 = "yet_another_use"(%0) : (tensor<?xf32>) -> (tensor<?xf32>)
+  //
+  // In this example, "some_op" will be cloned into the dispatch region and the
+  // OpOperand of "another_op" will be replaced:
+  //
+  // %0 = "some_op"() : () -> (tensor<?xf32>)
+  // %r = flow.dispatch.region -> (tensor<?xf32>{%d0}) {
+  //   %0_clone = "some_op"() : () -> (tensor<?xf32>)
+  //   %1 = "another_op"(%0_clone) : (tensor<?xf32>) -> (tensor<?xf32>)
+  //   flow.return %1 : tensor<?xf32>
+  // }
+  // %2 = "yet_another_use"(%0) : (tensor<?xf32>) -> (tensor<?xf32>)
+  //
+  // Note: If it weren't for "yet_another_use", the original "some_op" could
+  // fold away.
+  //
+  // Note: If there were no "another_op" in the dispatch region, the IR would
+  // not be modified and this function would fail.
+  if (domInfo.properlyDominates(target, regionOp)) {
+    SmallVector<OpOperand *> consumers;
+    for (OpOperand &use : target->getUses())
```

Comment:
I dont think you need this. If you are cloning the operation, then you are necessarily duplicating it within all regions that use it. So target dominating the region is all that should be needed. 

Another point to consider `properlyDominates` can be potentially expensive. Everytime you insert/modify an operation into a basic block, it invalidates the precomputed order that dominator information relies on. If you query dominance again, it will recompute the block order which requires traversal of the entire block. It is not unusual to have the entire program in the block. So this could be potentially O(n^2).